src/backend/commands/createas.c

```diff
@@ -958,11 +891,128 @@ intorel_destroy(DestReceiver *self)
 	pfree(self);
 }
 
+/*
+ * CreateIvmTriggersOnBaseTables -- create IVM triggers on all base tables
+ */
+void
+CreateIvmTriggersOnBaseTables(Query *qry, Oid matviewOid, bool is_create)
+{
+	Relids	relids = NULL;
+	bool	ex_lock = false;
+	Index	first_rtindex = is_create ? 1 : PRS2_NEW_VARNO + 1;
+	RangeTblEntry *rte;
+
+	/* Immediately return if we don't have any base tables. */
+	if (list_length(qry->rtable) < first_rtindex)
+		return;
+
+	/*
+	 * If the view has more than one base tables, we need an exclusive lock
+	 * on the view so that the view would be maintained serially to avoid
+	 * the inconsistency that occurs when two base tables are modified in
+	 * concurrent transactions. However, if the view has only one table,
+	 * we can use a weaker lock.
+	 *
+	 * The type of lock should be determined here, because if we check the
+	 * view definition at maintenance time, we need to acquire a weaker lock,
+	 * and upgrading the lock level after this increases probability of
+	 * deadlock.
+	 */
+
+	rte = list_nth(qry->rtable, first_rtindex - 1);
+	if (list_length(qry->rtable) > first_rtindex ||
```

Comment:
Maybe this patch can't support self join case. Is my understanding correct?
If query definition has a self join, rtable is two or more.

Reply:
The code is correct because if we have a self join, we will need an exclusive lock. Recall that we can regard same tables in the query as separated tables whose contents are same. Suppose a join view V = R*R and tuples t1 & t2 are inserted into R in concurrent transactions Tx1, Tx2 respectively. We can get t1*R and R*t2 in each transaction, we will fail to get t1*t2 without the exclusive lock.

Reply:
Thnak you! I understand this reason.